ext/ManoptLRUCacheExt.jl

```diff
@@ -0,0 +1,174 @@
+module ManoptLRUCacheExt
+
+if isdefined(Base, :get_extension)
+    using Manopt
+    using Manopt: AbstractDecoratedManifoldObjective, get_cost_and_gradient
+    import Manopt: get_cost, get_gradient_function, get_gradient, get_gradient!
+    using ManifoldsBase
+    using LRUCache
+else
+    # imports need to be relative for Requires.jl-based workflows:
+    # https://github.com/JuliaArrays/ArrayInterface.jl/pull/387
+    using ..Manopt
+    using ..Manopt: AbstractDecoratedManifoldObjective, get_cost_and_gradient
+    import ..Manopt: get_cost, get_gradient_function, get_gradient, get_gradient!
+    using ..ManifoldsBase
+    using ..LRUCache
+end
+
+@doc raw"""
+    LRUCacheObjective{E,P,O<:AbstractManifoldObjective{<:E},C<:NamedTuple{}} <: AbstractDecoratedManifoldObjective{E,P}
+
+Create a cache for an objective, based on a `NamedTuple` that stores `LRUCaches` for
+
+# Constructor
+
+    LRUCacheObjective(M, o::AbstractManifoldObjective, caches::Vector{Symbol}; kwargs...)
+
+Create a cache for the [`AbstractManifoldObjective`](@ref) where the Symbols in `caches` indicate,
+which function evaluations to cache.
+
+# Keyword Arguments
+* `p`           - (`rand(M)`) the type of the keys to be used in the caches. Defaults to the default representation on `M`.
+* `X`           - (`zero_vector(M,p)`) the type of values to be cached for gradient and Hessian calls.
+* `cache`       - (`[:Cost]`) a vector of symbols indicating which function calls should be cached.
+* `cache_size`  - (`10`) number of (least recently used) calls to cache
+* `cache_sizes` - a named tuple or dictionary specifying the sizes individually for each cache.
+"""
+struct LRUCacheObjective{E,P,O<:AbstractManifoldObjective{<:E},C<:NamedTuple{}} <:
+       AbstractDecoratedManifoldObjective{E,P}
+    objective::O
+    cache::C
+end
+function LRUCacheObjective(
+    M::AbstractManifold,
+    objective::O,
+    caches::AbstractVector{<:Symbol}=[:Cost];
+    p::P=rand(M),
+    v::R=get_cost(M, objective, p),
+    X::T=zero_vector(M, p),
+    cache_size=10,
+    cache_sizes=Dict{Symbol,Int}(),
+) where {E,O<:AbstractManifoldObjective{E},R,P,T}
+    c = init_caches(caches; p=p, v=v, X=X, cache_size=cache_size, cache_sizes=cache_sizes)
+    return LRUCacheObjective{E,O,O,typeof(c)}(objective, c)
+end
+function LRUCacheObjective(
```

Comment:
Would it maybe be reasonable to define this constructor `(m, args...; kwargs...) in `Manopt.jl` directly to be able to export it and inform the user that he needs to load LRU to use this?
That way if you get code that uses LRU but you have not loaded it upfront (because you only copied parts and not the using upfront)?

Something similar could be done for the solvers – and cool that you even spotted Chambolle-Pock.

Reply:
Cool. Will do that because it also makes the constructor “more visible” - so if you use autocomplete or something you can discover it and then you learn what you have to load to use it :)

Reply:
even better, with that constructor still being defined, the `objective_cache_factory` can also stay as before I think.

Reply:
Hm, but if it is already a function that throws an error (on a different unspecified signature) in Manopt I can not define a new function in the extension due to

```
ERROR: LoadError: cannot assign a value to imported variable Manopt.LRUCacheObjective from module ManoptLRUCacheExt
Stacktrace:
```

So an extension should not define new stuff but can also not define new signatures for old stuff? What am I missing here?

Reply:
So if I define a function with an error with the name of the constructor – I can not define the struct afterwards – if I do _not_ define the function – leading the extension does not make `LRUCacheObjective` available neither explorted not within `Manopt.jl`: Maybe this is all not possible? At least I do not see how structs and extensions presupposed to work together.

Reply:
> So an extension should not define new stuff but can also not define new signatures for old stuff? What am I missing here?

You can add methods to existing functions (and constructors to existing types) in package extensions. I'd have to take a look at what exactly you did in `ManoptLRUCacheExt` to be able to tell what could be wrong.

> So if I define a function with an error with the name of the constructor – I can not define the struct afterwards – if I do _not_ define the function – leading the extension does not make `LRUCacheObjective` available neither explorted not within `Manopt.jl`

If a method in Manopt uses a type, that type cannot be defined in an extension. If a type is defined in an extension, it (effectively) can only be instantiated by methods from said extension. But they can be methods of functions defined elsewhere.

Reply:
Hm, since the exams I am supposed to grade today. are still not scanned, I will twiddle around with this a bit, but I am not yet convinced that this will be any useful for any of the extensions we for now introduced, since it seems one should actually never define structs (unless for use only within the extension) within an extension.

For now we merely do the exact opposite.

Reply:
Yes, one big limitation is that structs defined in extensions can only be instantiated by methods from that extensions. As demonstrated, it can be worked around by either defining types in the main module or by making "factory"-like functions that instantiate types defined in extensions. So I don't see it as a big limitation, and maybe even a small win: it results in more easily extensible code. Interface enforced by that rule can be more easily adapted to a new library.

Reply:
For example, if someone made a "MyAwesomeCache.jl" library, with this structure it could be very easily used by code that was designed to use LRUCache.jl. If you tie a struct more closely to a cache library and let people instantiate it directly, more adjustments would be needed.

Reply:
I already switched to a name `CachedManifoldObjective` that might be able to work with arbitrary caches and only the LRU initialisation is done in the extension now.

Reply:
The newest commit (twiddely....) is an approach that might work. I will now check the other extensions before checking the tests again. I think there should sometimes be a reasonable error-default.

Reply:
> I think there should sometimes be a reasonable error-default.

I think error hints should work fine as a replacement: https://github.com/JuliaManifolds/Manifolds.jl/pull/607/files#diff-22f0e186bf4bcece078fec1aa646b8e18e436bfcf26138178f0a5c43c5dcf571R494 ?

Reply:
I lost about an hour trying to figure out these hints – I am not able to do that. But tests are back working and I hope my current (just the most recent) commit does something reasonable? Currently the new Cache init-function dispatches on the cache type and the default type (`Nothing`) returns an error. Loading LRU defines a more specialised function that even sets LRU as the default type.